### ISSUE TITLE
[AI] Infra: GPU Hang / Reset (1 job)

### ISSUE THREAD

**Issue by dhonnappa-amd** [dhonnappa-amd/build-dashboards-testing#6066]
## AI Analysis — Infrastructure Issue: GPU Hang / Reset

**1 affected job**

| Pipeline/Build | Queue | Job Link |
|---------------|-------|----------|
| amd-ci/7079 | `amd_mi355_1` | [View job](https://buildkite.com/vllm/amd-ci/builds/7079#019d3cc7-822b-4a26-9900-224f926541b4) |

---

### Summary

This build was testing the vLLM project's Mixture of Experts (MOE) kernel implementations on AMD ROCm GPUs. The job successfully completed the checkout phase, loaded the Docker CI image, performed a GPU reset, and began executing pytest on the MOE kernel test suite. Multiple tests were skipped (as expected for multi-GPU tests running on single GPU), and the test suite progressed to `test_modular_kernel_combinations_singlegpu`. This test successfully initialized with RMS normalization configuration using native operations. However, approximately 58 seconds into the kernel execution phase, the Python interpreter suffered a fatal segmentation fault. The crash occurred during torch multiprocessing spawn while the test was running Triton-based expert kernels with specific parameters (2048-1024-32 dimensions, dtype0, no quantization). The segmentation fault (exit code 139/SIGSEGV) indicates a memory access violation during GPU kernel execution, most likely due to a GPU hang, invalid memory access in the ROCm/HIP runtime, or a bug in the Triton kernel compilation for AMD GPUs. This is a genuine infrastructure failure, not a test logic error.

---

### Root Cause Analysis

Walking through the log chronologically: The job started normally with git checkout, Docker image loading, and GPU reset all succeeding. ROCm info showed the system was properly configured with ROCk module 6.16.6 and HSA runtime 1.15. The pytest suite began executing MOE kernel tests, correctly skipping multi-GPU tests. At 07:05:16, the `test_modular_kernel_combinations_singlegpu` test started and logged successful IR op priority initialization. Then at 07:06:14 (58 seconds later), the Python interpreter crashed with a segmentation fault. The stack trace shows the crash occurred in the main thread during `torch.multiprocessing.spawn` while waiting for worker processes, indicating a child process (likely running GPU kernels) crashed and caused the parent to segfault. The FIRST UNRECOVERABLE error is the segmentation fault at 07:06:14 - there were no prior errors, no retries, no transient failures. This is a hard crash during GPU kernel execution, characteristic of GPU hangs, memory corruption, or ROCm runtime issues on AMD hardware.

---

### Detected Events (all affected jobs)

**amd-ci/7079** (queue: `amd_mi355_1`):
- `gpu_hang`: Fatal Python error: Segmentation fault

Thread 0x0000705d924dc000 (most recent call first):
  File "/vllm-workspace/tests/kernels/moe/test_modular_kernel_combinations.py", line 376 in test_modular_kernel_combinations_singlegpu
  - _The Python interpreter crashed with a segmentation fault during GPU kernel execution in the MOE (Mixture of Experts) test. The stack trace shows the crash occurred during torch multiprocessing spawn while running modular kernel combinations test on a single GPU. This is a fatal GPU-related crash, not a test assertion failure._
- `gpu_hang`: /bin/bash: line 1:     8 Segmentation fault      pytest -v -s kernels/moe --ignore=kernels/moe/test_modular_oai_triton_moe.py --shard-id=$BUILDKITE_PARALLEL_JOB --num-shards=$BUILDKITE_PARALLEL_JOB_COUNT
...
Error: The command exited with status 139
  - _Exit status 139 (128 + 11) indicates the process was terminated by SIGSEGV (signal 11), confirming a segmentation fault. This is the shell's confirmation that the pytest process crashed due to a memory access violation, almost certainly during GPU kernel execution._
- `test_execution`: kernels/moe/test_modular_kernel_combinations.py::test_modular_kernel_combinations_singlegpu[2048-1024-32-dtype0-None-MoEPrepareAndFinalizeNoDPEPModular-TritonExperts-1] INFO 03-30 07:05:16 [kernel.py:196] Final IR op priority after setting platform defaults: IrOpPriorityConfig(rms_norm=['native'])
Fatal Python error: Segmentation fault
  - _The test was actively running and had successfully initialized the IR op priority configuration for RMS normalization. The crash occurred approximately 58 seconds later (07:05:16 to 07:06:14) during actual kernel execution, suggesting a GPU compute issue rather than a setup problem._

---

#### Immediate Fix

1. Check dmesg/kernel logs for amdgpu driver errors or GPU reset messages. 2. Verify ROCm version compatibility with the Triton kernels being tested. 3. Re-run the specific failing test in isolation: `pytest -v -s tests/kernels/moe/test_modular_kernel_combinations.py::test_modular_kernel_combinations_singlegpu[2048-1024-32-dtype0-None-MoEPrepareAndFinalizeNoDPEPModular-TritonExperts-1]` to confirm reproducibility. 4. If reproducible, try with smaller batch sizes or different kernel configurations to isolate the trigger.

#### Suggested Code Change

```
In `tests/kernels/moe/test_modular_kernel_combinations.py` around line 376, add GPU memory checks and error handling:

python
# Before parallel_launch_with_config call
import torch
if torch.cuda.is_available():
    torch.cuda.synchronize()
    torch.cuda.empty_cache()

try:
    parallel_launch_with_config(...)
except RuntimeError as e:
    if 'HIP' in str(e) or 'ROCm' in str(e):
        pytest.skip(f"GPU runtime error: {e}")
    raise


Also investigate `vllm/kernels/moe/modular_kernel_tools/parallel_utils.py` line 120 to add timeout and crash detection for spawned processes.
```

#### Prevention

1. Add pre-test GPU health checks in `.buildkite/scripts/hardware_ci/run-amd-test.sh` to verify GPU is responsive before running tests. 2. Implement per-test timeouts (e.g., 2 minutes) to catch hanging GPU kernels before they cause segfaults. 3. Add ROCm/HIP error checking wrapper around all GPU kernel launches in the test suite. 4. Enable core dumps in CI environment to capture crash details: `ulimit -c unlimited` before pytest. 5. Add telemetry to track which specific kernel configurations cause crashes on AMD GPUs.

#### Workaround

Temporarily skip this specific test configuration by adding to `tests/kernels/moe/test_modular_kernel_combinations.py`:

python
@pytest.mark.skipif(
    os.environ.get('BUILDKITE_AGENT_META_DATA_QUEUE') == 'amd',
    reason='Segfault on AMD GPUs with TritonExperts - tracking issue #XXXXX'
)
def test_modular_kernel_combinations_singlegpu(...):

Or reduce test matrix by excluding the problematic dtype/config combination until the root cause in Triton/ROCm is identified.

---

_This issue was automatically created by the AI log analysis pipeline._
